src/box/box.cc

```diff
@@ -681,6 +691,191 @@ box_check_uri(const char *source, const char *option_name)
 	return 0;
 }
 
+/**
+ * Check that string which contain one or several URIs with options
+ * does not start or end with one of the listed delimiters.
+ * URI should not contain delimiters, which are used to separate
+ * different URIs or URI and it's options. This is important because
+ * `strtok` and `strtok_r` functions skip delimiters passed them as
+ * arguments. So if user pass some strange URI like `,/???` we unable
+ * to catch error, because `strtok_r` skip ',' and `?` symbols.
+ * @param source string containing one or seveal URIs with options.
+ *               It supposed to be NULL-terminated.
+ */
+static inline int
+box_check_uri_delimiters(const char *source)
+{
+	static const char *uri_delimiters = ", ?=&;\0";
+	static const uint32_t uri_delimiters_size = 7;
+	size_t source_len = strlen(source);
+	source_len = (source_len != 0 ? source_len : source_len + 1);
+
+	for (uint32_t i = 0; i < uri_delimiters_size; i++) {
+		if (source[0] == uri_delimiters[i] ||
+		    source[source_len - 1] == uri_delimiters[i]) {
+			diag_set(ClientError, ER_CFG, "listen",
+				 "expected host:service or /unix.socket");
+			return -1;
+		}
+	}
+	return 0;
+}
+
+static inline int
+box_check_end_of_token(const char *delimiter, char **saveptr,
+		       const char *error)
+{
+	if (strtok_r(NULL, delimiter, saveptr) != NULL) {
+		diag_set(ClientError, ER_CFG, "listen", error);
+		return -1;
+	}
+	return 0;
+}
+
+static inline enum uri_option_key
+parse_uri_option_key(const char *opt)
+{
+	if (strncmp(opt, "transport", strlen("transport")) == 0)
+		return URI_OPTION_KEY_TRANPORT;
+	return URI_OPTION_KEY_UNKNOWN;
+}
+
+
+static inline enum uri_option_value
+parse_uri_option_value(const enum uri_option_key key, const char *val)
+{
+	switch(key) {
```

Comment:
I guess parsing of an options should be independent from key - check that key corresponds to a proper value should be done in a caller.